starknet_py/net/account/account.py

```diff
@@ -0,0 +1,394 @@
+from starknet_py.utils.typed_data import TypedData as TypedDataDataclass
+
+
+class Account(BaseAccount):
+    """
+    Default Account implementation.
+    """
+
+    def __init__(
+        self,
+        address: AddressRepresentation,
+        client: Client,
+        signer: Optional[BaseSigner] = None,
+        key_pair: Optional[KeyPair] = None,
+        chain: Optional[StarknetChainId] = None,
+    ):
+        # pylint: disable=too-many-arguments
+        if chain is None and signer is None:
+            raise ValueError("One of chain or signer must be provided")
+
+        self.address = parse_address(address)
+        self._client = client
+
+        if signer is None:
+            if key_pair is None:
+                raise ValueError(
+                    "Either a signer or a key_pair must be provided in AccountClient constructor"
+                )
+
+            chain = chain_from_network(net=client.net, chain=chain)
+            signer = StarkCurveSigner(
+                account_address=self.address, key_pair=key_pair, chain_id=chain
+            )
+        self.signer: BaseSigner = signer
+        self._version = 1
+
+    @property
+    def client(self) -> Client:
+        return self._client
+
+    @property
+    def supported_tx_version(self) -> int:
+        return self._version
+
+    def _get_default_token_address(self) -> str:
+        if self._client.net not in [TESTNET, MAINNET]:
+            raise ValueError(
+                "Token_address must be specified when using a custom net address"
+            )
+
+        return FEE_CONTRACT_ADDRESS
+
+    async def _get_max_fee(
+        self,
+        transaction: Union[InvokeFunction, Declare, DeployAccount],
+        max_fee: Optional[int] = None,
+        auto_estimate: bool = False,
+    ) -> int:
+        if auto_estimate and max_fee is not None:
+            raise ValueError(
+                "Max_fee and auto_estimate are exclusive and cannot be provided at the same time."
+            )
+
+        if auto_estimate:
+            estimate_fee = await self._signed_estimate_fee(transaction)
+            max_fee = int(estimate_fee.overall_fee * 1.1)
+
+        if max_fee is None:
+            raise ValueError("Max_fee must be specified when invoking a transaction")
+
+        return max_fee
+
+    async def _prepare_invoke_function(
+        self,
+        calls: Calls,
+        max_fee: Optional[int] = None,
+        auto_estimate: bool = False,
+    ) -> InvokeFunction:
+        """
+        Takes calls and creates InvokeFunction from them
+
+        :param calls: Single call or list of calls
+        :param max_fee: Max amount of Wei to be paid when executing transaction
+        :param auto_estimate: Use automatic fee estimation, not recommend as it may lead to high costs
+        :return: InvokeFunction created from the calls (without the signature)
+        """
+        nonce = await self.get_nonce()
+
+        calldata_py = _merge_calls(ensure_iterable(calls))
```

Comment:
I find this a little confusing :/ `ensure_iterable` returns [calls] if calls are not a list, but maybe we could find a better name?

Reply:
You mean the `ensure_iterable` itself? I think the name is ok actually as it is exactly what it does, but I was probably the one who created it :)

Reply:
Maybe it was just me :? Resolved!